### PR TITLE
Catch exceptions while calling builder factories

### LINKED ISSUE #1557: Invalid builder flags reported as a bug in build_runner
* [x] Dart SDK Version (`dart --version`): 2.0.0-dev.60.0
* [x] What package(s) from this repo you are using, and the version (i.e. `build_runner 0.7.12`)
  * build 0.12.6
  * build_config 0.3.0
  * build_modules 0.2.2+6
  * build_resolvers 0.2.0+2
  * build_runner 0.8.10
  * build_runner_core 0.1.0
  * build_test 0.10.2+4
  * build_web_compilers 0.4.0+4
* [x] What builder(s) you are using (or writing yourself). Try to give a short summary of what they do.
  * https://github.com/dart-lang/angular/blob/master/angular/build.yaml#L12
* [x] Whether you are using Windows, MacOSX, or Linux (if applicable)
  * macOS

Adding an invalid flag (doesn't exist) to a builder, is reported as a bug in build_runner itself. To reproduce:

```
$ git clone git@github.com:dart-lang/angular.git
$ cd angular/examples/hello_world
```

Edit `build.yaml`:

```yaml
targets:
  $default:
    builders:
      # --- Add these lines --- 
      angular|angular:
        options:
          foo: True # The angular builder doesn't define this flag.
      # -----------------------
      build_web_compilers|entrypoint:
        generate_for:
           - web/main.dart
        release_options:
          dart2js_args:
            - --fast-startup
            - --minify
            - --trust-type-annotations
            - --trust-primitives
```

```
$ pub upgrade
$ pub run build_runner build
```

The output should be something like this:

```
[INFO] Generating build script completed, took 355ms
[SEVERE] angular|angular:

Invalid argument(s): Invalid arguments passed to the transformer: 
  - foo

You may be providing flags that are no longer valid or supported for AngularDart 5.x. See "compiler_flags.md" in the AngularDart repository for a list of supported flags.

You have hit a bug in build_runner
Please file an issue with reproduction steps at https://github.com/dart-lang/build/issues


builderFactory did not return a builder.
package:build_runner_core/src/package_graph/apply_builders.dart 180:30  new BuilderApplication.forBuilder.<fn>.<fn>
package:build_runner_core/src/package_graph/apply_builders.dart 320:27  _createBuildPhasesForBuilderInCycle.<fn>.<fn>
dart:core                                                               List.addAll
package:build_runner_core/src/package_graph/apply_builders.dart 269:7   createBuildPhases
package:build_runner_core/src/generate/build_impl.dart 97:29            BuildImpl.create
package:build_runner_core/src/generate/build_runner.dart 31:41          BuildRunner.create
package:build_runner/src/generate/build.dart 91:35                      build
package:build_runner/src/entrypoint/build.dart 28:24                    BuildCommand.run
package:args/command_runner.dart 194:27                                 CommandRunner.runCommand
package:args/command_runner.dart 109:29                                 CommandRunner.run.<fn>
dart:async                                                              new Future.sync
package:args/command_runner.dart 109:11                                 CommandRunner.run
package:build_runner/src/entrypoint/run.dart 22:31                      run
.dart_tool/build/entrypoint/build.dart 80:22                            main
```

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #1557

Prevents code which is out of our control (from builder authors) from
showing up as an apparent build_runner bug.

### CODE REVIEW

File: build_runner_core/lib/src/package_graph/apply_builders.dart [hunk 1]
```diff
@@ -175,9 +175,19 @@ class BuilderApplication {
...
-        var builder = _scopeLogSync(
-            () => builderFactory(optionsWithDefaults), new Logger(builderKey));
-        if (builder == null) throw 'builderFactory did not return a builder.';
+        final logger = new Logger(builderKey);
+        Builder builder;
+        try {
+          builder =
+              _scopeLogSync(() => builderFactory(optionsWithDefaults), logger);
+        } catch (e, st) {
+          logger.severe('Failed to instantiate builder.', e, st);
```
Comment (jakemac53) [inline build_runner_core/lib/src/package_graph/apply_builders.dart]: nit: log the package and options

Comment (jakemac53) [reply build_runner_core/lib/src/package_graph/apply_builders.dart]: also use `new JsonEncoder.withIndent('  ').convert(e.options.config)` for the options ;)

Comment (jakemac53) [reply build_runner_core/lib/src/package_graph/apply_builders.dart]: also note errors won't be thrown here its always just the null case and we already log the error with severe

Comment (natebosch) [reply build_runner_core/lib/src/package_graph/apply_builders.dart]: Done

**Pull request merged**